File: push_swap_prototype/internal/stack/op_swap.go
package stack

// swap swaps the first 2 elements at the top of a stack (exactly like C swap)
func swap(stk *Stack) {
	// C implementation checks if second element exists by checking if it's 0
	// stk->stack[next_down(stk, stk->top)] == 0
	if stk.CurrentSize() < 2 {
		return
	}
	
	// Get the second element position using NextDown like C implementation
	secondPos := stk.NextDown(stk.GetTop())
	
	// Check if second element exists (not 0) - exactly like C implementation
	if stk.GetStack()[secondPos] == 0 {
		return
	}
	
	// Swap the top two elements
	stk.SwapTop()
}

// Swap_a swaps the first 2 elements at the top of stack A (like C swap_a)
func Swap_a(data *PushSwapData) {
	swap(data.A)
	if data.WritingMode {
		saveOp(data, SA)
	}
}

// Swap_b swaps the first 2 elements at the top of stack B (like C swap_b)
func Swap_b(data *PushSwapData) {
	swap(data.B)
	if data.WritingMode {
		saveOp(data, SB)
	}
}

// Swap_ab swaps the first 2 elements at the top of both stacks (like C swap_ab)
func Swap_ab(data *PushSwapData) {
	swap(data.A)
	swap(data.B)
	if data.WritingMode {
		saveOp(data, SS)
	}
}

File: push_swap_prototype/lab/push_swap/src/op_swap.c

#include "models.h"

void	save_op(t_ps *data, t_op op);

int	next_down(t_stack *stk, int index);

void	swap(t_stack *stk)
{
	int	tmp;

	if (stk->stack[next_down(stk, stk->top)] == 0)
		return ;
	tmp = stk->stack[next_down(stk, stk->top)];
	stk->stack[next_down(stk, stk->top)] = stk->stack[stk->top];
	stk->stack[stk->top] = tmp;
}

void	swap_a(t_ps *data)
{
	swap(&data->a);
	if (data->writing_mode)
		save_op(data, sa);
}

void	swap_b(t_ps *data)
{
	swap(&data->b);
	if (data->writing_mode)
		save_op(data, sb);
}

void	swap_ab(t_ps *data)
{
	swap(&data->a);
	swap(&data->b);
	if (data->writing_mode)
		save_op(data, ss);
}